### PR TITLE
GoQuorum privacy private receipts

### PR DESCRIPTION
**Pull request by pinges**
- fix nonce problem
- remove not needed method on GoQuorumPrivateStorage
- make receipts work as they work in GoQuorum

Signed-off-by: Stefan Pingel <stefan.pingel@consensys.net>

### CODE REVIEW

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/goquorum/GoQuorumBlockProcessor.java [hunk 1]
```diff
@@ -99,71 +102,92 @@ public Result processBlock(
...
+      }
 
-        final TransactionReceipt privateTransactionReceipt =
-            transactionReceiptFactory.create(
-                transaction.getType(), result, privateWorldState, currentGasUsed);
-        privateStorageUpdater.putTransactionReceipt(
-            blockHeader.getHash(), transaction.getHash(), privateTransactionReceipt);
-      } else {
+      if (effectiveTransaction
+          != null) { // public transaction, or private transaction that we are party to
```
[the remark this answers is not in the document]
Comment (macfarla) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/goquorum/GoQuorumBlockProcessor.java:129]: transaction -> tx would make the comment fit on the same line

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/AbstractBlockProcessor.java [hunk 2]
```diff
@@ -67,26 +68,49 @@ TransactionReceipt create(
     private final boolean successful;
 
     private final List<TransactionReceipt> receipts;
+    private final Optional<List<TransactionReceipt>> privateReceipts;
```
Comment (lucassaldanha) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/AbstractBlockProcessor.java:71]: NIT: Do we really need to have this list as an optional? Can't we only check if it is empty instead?

**Review by lucassaldanha — APPROVED** [review]

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/goquorum/GoQuorumBlockProcessor.java [hunk 3]
```diff
@@ -99,71 +102,92 @@ public Result processBlock(
...
-        final TransactionProcessingResult publicResult =
-            TransactionProcessingResult.successful(
-                Collections.emptyList(),
-                0,
-                result.getGasRemaining(),
-                Bytes.EMPTY,
-                result.getValidationResult());
+        } catch (final EnclaveClientException e) { // private transaction but not party to it
+          effectiveTransaction = null;
+          // TODO-goquorum: we should probably still check the signature and the nonce
```
[the remark this answers is not in the document]
Comment (pinges) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/goquorum/GoQuorumBlockProcessor.java:120]: I'll add the code before we merge the PR!

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/goquorum/GoQuorumBlockProcessor.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (pinges) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/goquorum/GoQuorumBlockProcessor.java:129]: Done!

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/goquorum/GoQuorumBlockProcessor.java [hunk 4]
```diff
@@ -99,71 +102,92 @@ public Result processBlock(
...
+        }
+      } else { // private transaction we are not party to
         publicTxReceipts.add(
             transactionReceiptFactory.create(
-                transaction.getType(), result, publicWorldState, currentGasUsed));
+                transaction.getType(),
+                publicResultForWhenWeHaveAPrivateTransaction(transaction),
+                publicWorldState,
+                currentGasUsed));
+        privateTxReceipts.add(null);
```
[the remark this answers is not in the document]
Comment (pinges) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/goquorum/GoQuorumBlockProcessor.java:182]: I'm adding null so I replace the public receipt with the private one if it is not null. Could probably come up with a different way of storing the private ones that will allow me to replace the right public receipt ...

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/AbstractBlockProcessor.java (hunk 2 above)
Comment (pinges) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/AbstractBlockProcessor.java:71]: Done

**Pull request merged**